src/wallet/rpcwallet.cpp

```diff
@@ -3085,50 +3087,115 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
     }
-    int64_t account = params[0].get_int64();
-    if (account < 0 || account >= ZCASH_LEGACY_ACCOUNT) {
+
+    LOCK(pwalletMain->cs_wallet);
+
+    int64_t accountInt = params[0].get_int64();
+    if (accountInt < 0 || accountInt >= ZCASH_LEGACY_ACCOUNT) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid account number, must be 0 <= account <= (2^31)-2.");
     }
-    // TODO: Check that the account is known to the wallet.
-    std::vector<uint8_t> parsed_diversifier_index;
+    libzcash::AccountId account = accountInt;
+
+    std::optional<libzcash::diversifier_index_t> j = std::nullopt;
     if (params.size() >= 2) {
+        // TODO: Handle '*'
         if (params[1].getType() != UniValue::VNUM) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid diversifier index, must be an unsigned integer.");
         }
         auto parsed_diversifier_index_opt = ParseArbitraryInt(params[1].getValStr());
         if (!parsed_diversifier_index_opt.has_value()) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "diversifier index must be a decimal integer.");
         }
-        parsed_diversifier_index = parsed_diversifier_index_opt.value();
+        auto parsed_diversifier_index = parsed_diversifier_index_opt.value();
         if (parsed_diversifier_index.size() > ZC_DIVERSIFIER_SIZE) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "diversifier index is too large.");
         }
-    } else {
-        // TODO get next unused diversifier index from wallet
+        j = libzcash::diversifier_index_t(parsed_diversifier_index);
     }
-    // TODO:
-    // diversifier_t diversifier{};
-    // std::copy(parsed_diversifier_index.begin(), parsed_diversifier_index.end(), diversifier.begin());
-    UniValue pools(UniValue::VARR);
+
+    std::set<libzcash::ReceiverType> receivers;
     if (params.size() >= 3) {
-        pools = params[2].get_array();
+        const auto& pools = params[2].get_array();
         for (unsigned int i = 0; i < pools.size(); i++) {
             const std::string& p = pools[i].get_str();
-            if (!(p == "transparent" || p == "sapling" || p == "orchard")) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "pool arguments must be \"transparent\", \"sapling\", or \"orchard\"");
+            if (p == "transparent") {
+                receivers.insert(ReceiverType::P2PKH);
+            } else if (p == "sapling") {
+                receivers.insert(ReceiverType::Sapling);
+            } else {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "pool arguments must be \"transparent\", or \"sapling\"");
             }
         }
     } else {
-        // default is all
-        pools.push_back("transparent");
-        pools.push_back("sapling");
-        pools.push_back("orchard");
+        // Default is the best and second-best shielded pools, and the transparent pool.
+        receivers = {ReceiverType::P2PKH, ReceiverType::Sapling};
     }
+
+    EnsureWalletIsUnlocked();
```

Comment:
Okay, so the current implementation of the wallet needs to be modified so that we don't need to have the wallet unlocked here; I'll add a ticket to that effect. Essentially, we should cache the UFVKs when the wallet is unlocked. This will do for now though.

Reply:
We need the wallet unlocked for the transparent components. We _could_ gate this check on whether `receivers` contains `ReceiverType::P2PKH`.